torch/distributed/_shard/checkpoint/resharding.py

```diff
@@ -0,0 +1,290 @@
+
+        narrows.append(
+            (dim, offset_for_saved_tensor, offset_for_current_tensor, length)
+        )
+
+    return narrows
+
+
+def _get_sharded_tensor_element_size(tensor: ShardedTensor) -> int:
+    if len(tensor.local_shards()) > 0:
+        test_tensor = tensor.local_shards()[0].tensor
+    else:
+        dtype = tensor.metadata().tensor_properties.dtype
+        test_tensor = torch.empty((1,), dtype=dtype)
+
+    return test_tensor.element_size()
+
+
+def _compute_sharded_tensor_md(
+    tensor: ShardedTensor,
+    shard_to_storage_key: Dict[str, str]
+) -> ShardedTensorStorageMetadata:
+    smd = []
+    for shard_md in tensor.metadata().shards_metadata:
+        shard_storage_key = shard_to_storage_key[_get_shard_key(shard_md)]
+
+        shard_size = 1
+        for d in shard_md.shard_sizes:
+            shard_size *= d
+
+        # not particularly great
+        storage_size = shard_size * _get_sharded_tensor_element_size(tensor)
+
+        one_smd = ShardStorageMetadata(
+            shard_metadata=shard_md,
+            storage_key=shard_storage_key,
+            length=storage_size,
+        )
+        smd.append(one_smd)
+
+    return ShardedTensorStorageMetadata(
+        tensor_metadata=tensor.metadata(),
+        storage_metadata=smd,
+    )
+
+
+def _get_shard_key(shard: ShardMetadata) -> str:
+    """
+    Compute an unique key for a shard.
+
+    This key is unique vis-a-vis other shard of the owning ShardedTensor
+    """
+    return "_".join([str(i) for i in shard.shard_offsets])
+
+def _get_shard_storage_key(
+    tensor_storage_key: str,
+    shard: ShardMetadata,
+    storage_key_to_fqn: Dict[str, str]
+) -> str:
+    shard_key = f"{tensor_storage_key}{STORAGE_KEY_SEPARATOR}{_get_shard_key(shard)}"
+
+    return _create_storage_key(storage_key_to_fqn, shard_key)
+
+
+def _prepare_sharded_tensor_write(
+    sharded_tensor: ShardedTensor,
+    storage_key: str,
+    storage_key_to_fqn: Dict[str, str]
+) -> Tuple[List[TensorWriteRequest], ShardedTensorStorageMetadata]:
+    """
+    Prepare sharded tensor write.
+
+    Args:
+        sharded_tensor: The sharded tensor to persist.
+        storage_key: The identifier for `sharded_tensor`.
+        storage_key_to_fqn: dict used to produce storage keys
+
+    Returns:
+        Write requests for persisting the sharded tensor, and metadata
+        describing the persisted sharded tensor.
+
+    NB `storage_key` is used to compose the key names of the local shards.
+
+    """
+    write_requests = []
+    shard_to_storage_key: Dict[str, str] = dict()
+
+    for shard_md in sharded_tensor.metadata().shards_metadata:
+        shard_storage_key = _get_shard_storage_key(storage_key, shard_md, storage_key_to_fqn)
+        shard_to_storage_key[_get_shard_key(shard_md)] = shard_storage_key
+
+    for shard in sharded_tensor.local_shards():
+        tensor = shard.tensor.detach()
+        shard_storage_key = shard_to_storage_key[_get_shard_key(shard.metadata)]
```

Comment:
Why do we need the loop above? Can't we just call `_get_shard_storage_key` here and eliminate the loop above?